Refactored + some new features

domchom opened the pull request:
The code has been refactored into separate files. Several features have been added, but we may not want to include all of them (I'll try to remember them all):

- Kymograph analysis: Similar analysis as the original "standard" version, but processes kymographs instead.
- Fixed an issue for instances of a very small time shift (~3 frames or less) where the code would previously select the CCF peak second closest to zero, thereby adding an entire period to the shift value. This was fixed by subtracting an entire period from the shift if the shift is greater than 60% of the period.
- Fixed a simple issue where the script would fail if no peaks were found in some signals. The solution was to assign some of the peak property values to NaN.
- Automatically calculate statistics related to time (e.g., shift, period, etc.) based on seconds instead of frames, if the movie itself has frame interval metadata. The script will give the user a warning if there is no metadata and will assume a frame interval of 1 (This might be something that we do not want in the official version, but I got tired of always calculating those numbers afterward).
- A new stat called "peak offset": We had some waves that are not totally sinusoidal but are more of a short ramp-off signal and sharp drop-off. One method we thought of to quantify this was to measure the distance between the location of the peak signal and the location of the midpoint of the adjacent troughs.
- Not a new stat, but sometimes useful: % phase shift. This is simply the time-shift divided by the period.

Quality of life updates:

- The script now saves the individual values for the CCF plots. This is useful for using other plotting software for figures.
- For a similar reason, it now saves all the mean values into individual CSV files, making it much easier to copy and paste into other graphing software.

There are also some remnants of code related to an optional calculation of wave speed via a user drawing lines on the kymographs using a napari window. I was able to get the code to work in a separate workspace, but there was a weird error in this workspace related to loading the napari window that I was never able to figure out, so I gave up. We can easily delete the code remnants, though.

I think that is everything.

zacswider commented on tests/test_image_props/test_standard_rolling_image_props.py:
```diff
@@ -0,0 +1,24 @@
+@pytest.fixture
+def default_filepaths():
+    return [
+        '/Users/domchom/Documents/GitHub/ZS_wave_analysis/tests/assets/standard/1_Group1.tif',
```
you can change all of these to relative paths (ie delete `/Users/domchom/Documents/GitHub/ZS_wave_analysis/`) since the user should be running `pytest` from the root directory and obviously won't have the upstream file path in their system

zacswider replied:
Probably because `.csv` is in the .gitignore file, oops! you can use `git add -f path/to/your/file.csv` to force add a .csv file

zacswider commented on tests/test_workflows/test_rolling_workflow.py:
```diff
@@ -0,0 +1,36 @@
+
+def test_rolling_workflow(default_log_params):
+    # load csv
+    known_results = pd.read_csv('tests/assets/rolling/1_Group2_summary.csv')
```
it looks like these `.csv` files didn't get committed to the repo, so these tests fail

zacswider replied:
Probably because .csv is in the .gitignore file, oops! you can use git add -f path/to/your/file.csv to force add a .csv file

zacswider commented on waveanalysis/__main__.py:
```diff
@@ -1,41 +1,37 @@
-from waveanalysis.waveanalysismods.processor import TotalSignalProcessor, RollingSignalProcessor
-
-np.seterr(divide='ignore', invalid='ignore')
+import housekeeping.housekeeping_functions as hf
+from custom_gui import BaseGUI, RollingGUI, KymographGUI
+from data_workflows.combined_workflow import combined_workflow
+from data_workflows.rolling_workflow import rolling_workflow
```
these  to be `import waveanalysis.housekeeping...` or `from waveanalysis...` because (for example) `housekeeping` won't be importable unless the user clones the repo and runs from within `waveanalys`

zacswider commented on waveanalysis/data_workflows/rolling_workflow.py:
```diff
@@ -0,0 +1,262 @@
+from tqdm import tqdm
+from typing import Any
+import scipy.signal as sig
+import plotting as pt
+import signal_processing as sp
+import housekeeping.housekeeping_functions as hf
+
+from image_props.image_bin_calc import create_multi_frame_bin_array
+from image_props.image_to_np_arrays import tiff_to_np_array_multi_frame
+from image_props.image_properties import get_multi_frame_properties
+from summarize_save.summarize_images import summarize_image, combine_stats_rolling
```
these also need to be prefixed with `wavenalysis.`

zacswider commented on waveanalysis/data_workflows/combined_workflow.py:
```diff
@@ -0,0 +1,362 @@
+import pandas as pd
+from tqdm import tqdm
+from typing import Any
+import plotting as pt
+import signal_processing as sp
+import housekeeping.housekeeping_functions as hf 
+
+from image_props.image_bin_calc import create_multi_frame_bin_array, create_kymo_bin_array
+from image_props.image_to_np_arrays import tiff_to_np_array_multi_frame, tiff_to_np_array_single_frame
+from image_props.image_properties import get_multi_frame_properties, get_single_frame_properties
+from summarize_save.save_stats import save_parameter_means_to_csv, get_mean_CCF_values, get_indv_CCF_values, save_ccf_values_to_csv
+from summarize_save.summarize_images import summarize_image, combine_stats_for_image_kymo_standard
```
these also need to be prefixed with `wavenalysis.`

zacswider commented on waveanalysis/data_workflows/combined_workflow.py:
```diff
@@ -0,0 +1,362 @@
+        if group_names != ['']:
+            # generate comparisons between each group
+            mean_parameter_figs = pt.generate_group_comparison(summary_df = summary_df, log_params = log_params)
+            group_plots_save_path = os.path.join(main_save_path, "!group_comparison_graphs")
+            os.makedirs(group_plots_save_path, exist_ok=True)
+            hf.save_plots(mean_parameter_figs, group_plots_save_path)
+
+            # save the means each parameter for the attributes to make them easier to work with 
+            parameter_tables_dict = save_parameter_means_to_csv(summary_df=summary_df,group_names=group_names)
+            mean_measurements_save_path = os.path.join(main_save_path, "!mean_parameter_measurements")
```
did you mean to put an `!` at the beginning of the output assets?

zacswider commented on waveanalysis/data_workflows/rolling_workflow.py:
```diff
@@ -0,0 +1,262 @@
+            summary_df = combine_stats_rolling(
+                img_props_dict=img_props_dict,
+                img_parameters_dict=img_parameters_dict,
+                indv_ccfs=indv_ccfs
```
`indv_ccfs` will be unbound if the dataset is only 1 channel. I ran into this because the sample dataset in `assets/data/test_data_rolling/` is 1 channel so you can use that for testing. You can probably make the argument to this function optional and then either pass `indv_ccfs` or None at runtime depending on the dataset `indv_ccfs=indv_ccfs if num_channels > 1 else None`

zacswider commented on waveanalysis/summarize_save/summarize_images.py:
```diff
@@ -0,0 +1,252 @@
+
+    # Extract image parameters from the dictionary
+    indv_periods = img_parameters_dict['Period']
+    indv_shifts = img_parameters_dict['Shift']
+    indv_peak_widths = img_parameters_dict['Peak Width']
+    indv_phase_shifts = img_parameters_dict['% Phase Shift']
```
`Shift` and `% Phase Shift` will both fail with `KeyError` if the dataset is 1 channel.

zacswider commented on waveanalysis/summarize_save/summarize_images.py:
```diff
@@ -0,0 +1,252 @@
+        submovie_summary = {'Submovie': submovie + 1}
+
+        # Calculate percentage of no shifts for each channel combination
+        if num_channels > 1:
```
probably easiest to unpack `indv_shifts` and `indv_phase_shifts` inside this block since they won't exist otherwise

zacswider reviewed (changes requested):
Sorry this took so long to look through. The imports should be addressed before merging because otherwise the install instructions in the readme don't work. The comments in the rolling analysis should be pretty straightforward so that's also worth doing before merging. I can't get all your tests to pass because it looks like not all the test data was committed to the repo. That would be a nice thing to have for future developers but definitely not as urgent as the first two things. After making some local changes to fix 1&2 I was able to get all 3 analysis types to produce output that looks as expected to me, so that's wonderful!

One more strange finding - if click `Launch kymograph analysis` in the gui and then click `Select folder` I sometimes get the file explorer window to open and I sometimes get a segmentation fault... huge wtf. Presumably this is an issue with tkinter as the kymograph gui is basically exactly the same as the rolling gui, which has no issues. Everything works fine if I paste in a file path manually. IDK what to say except maybe have someone else test it too?

domchom commented:
I have updated the code with the requested changes. Specifically:

- Added "waveanalysis." to many of the imports. I had to move the "main.py" file outside of the "waveanalysis" folder for it to work for me now. I'm not sure why it doesn't work when it is in that folder. Additionally, I'm uncertain if I need to move the "setup.py" file outside of that folder as well.
- Uploaded the required .csv/.json/.pkl files needed for testing. All tests should pass now (I hope). However, the file names are still disorganized and should probably be revisited.
- Modified the script so that it continues running even if there is an issue with one of the movies, sending a warning to the user that some movies were skipped.
- Identified and fixed a few additional issues for movies that do not have waves at all (i.e., control cells with probe only).
- Added small code chunks to prevent an output folder from being generated when automatically running the test (something that should have been done earlier).

Regarding the kymograph select folder bug, I initially encountered the same error a few times, but it then seemed to go away. I have since tested it on three other people's computers and have not run into the issue, so I am also unsure what is happening.

zacswider reviewed (approved):
looks good!
